vsintegration/src/FSharp.Editor/FSharpColorizationService.fs

```diff
@@ -0,0 +1,143 @@
+
+        for i = scanStartLine to scanEndLine do
+            cancellationToken.ThrowIfCancellationRequested()
+            let currentLine = sourceText.Lines.Item(i)
+            scanSourceLine(currentLine, runningLexState)
+
+        SourceTextColorizationData(result)
+        
+    static let classifySourceTextAsync(sourceText: SourceText, fileName: Option<string>, textSpan: TextSpan, result: List<ClassifiedSpan>, cancellationToken: CancellationToken) =
+        Task.Run(fun () ->
+            try
+                let classificationData = FSharpColorizationService.GetColorizationData(sourceText, textSpan, fileName, [], cancellationToken)
+                result.AddRange(classificationData.Tokens |> Seq.filter(fun token -> textSpan.Start <= token.TextSpan.Start && token.TextSpan.End <= textSpan.End))
+            with ex -> 
+                Assert.Exception(ex)
+                reraise()  
+        )
+
+    interface IEditorClassificationService with
+        
+        member this.AddLexicalClassifications(text: SourceText, textSpan: TextSpan, result: List<ClassifiedSpan>, cancellationToken: CancellationToken) =
+            classifySourceTextAsync(text, None, textSpan, result, cancellationToken).Wait(cancellationToken)
+        
+        member this.AddSyntacticClassificationsAsync(document: Document, textSpan: TextSpan, result: List<ClassifiedSpan>, cancellationToken: CancellationToken) =
+            let sourceText = document.GetTextAsync(cancellationToken).ConfigureAwait(false).GetAwaiter().GetResult()
+            classifySourceTextAsync(sourceText, Some(document.Name), textSpan, result, cancellationToken)
+
+        member this.AddSemanticClassificationsAsync(document: Document, textSpan: TextSpan, result: List<ClassifiedSpan>, cancellationToken: CancellationToken) =
+            let sourceText = document.GetTextAsync(cancellationToken).ConfigureAwait(false).GetAwaiter().GetResult()
+            classifySourceTextAsync(sourceText, Some(document.Name), textSpan, result, cancellationToken)
+
+        member this.AdjustStaleClassification(text: SourceText, classifiedSpan: ClassifiedSpan) : ClassifiedSpan =
+            let result = new List<ClassifiedSpan>()
+            classifySourceTextAsync(text, None, classifiedSpan.TextSpan, result, CancellationToken.None).Wait()
+            if result.Any() then
+                result.First()
+            else
+                new ClassifiedSpan(ClassificationTypeNames.WhiteSpace, classifiedSpan.TextSpan)
+
+    // Helper function to proxy Roslyn types to tests
+    static member GetColorizationData(sourceText: SourceText, textSpan: TextSpan, fileName: Option<string>, defines: string list, cancellationToken: CancellationToken) : SourceTextColorizationData =
+        colorizationDataCache.GetValue((sourceText, textSpan, fileName), fun key -> scanSourceText(sourceText, textSpan, fileName, defines, cancellationToken))
```

Comment:
About the comment, it could be replaced in classifySourceTextAsync() if not the need for tests.
About `fileName`, my understanding is that `SourceText` is not tied to a certain document.
We don't want to colorize two files having the same text the same, because it might depend on #defines supplied by the compiler (example).